config/common.go

```diff
@@ -140,3 +156,75 @@ func (cs *commonStore) validate(cfg *model.Config) error {
 
 	return nil
 }
+
+// mergeConfig merges two configs together. The receiver's values are overwritten with the patch's
+// values except when the patch's values are nil.
+func (cs *commonStore) mergeConfig(patch *model.Config) (*model.Config, error) {
+	ret, err := utils.Merge(cs.config, patch)
+	if err != nil {
+		return nil, err
+	}
+	retC := ret.(model.Config)
+	return &retC, nil
+}
+
+// removeEnvOverrides takes the newCfg provided and adds information stored in the commonStore,
+// then delegates the task to the removeEnvOverrides function
+func (cs *commonStore) removeEnvOverrides(newCfg *model.Config) *model.Config {
+	return removeEnvOverrides(newCfg, cs.configWithoutOverrides, cs.environmentOverrides)
+}
+
+// removeEnvOverrides returns a new config without the current environment overrides.
+// If a config variable has an environment override, that variable is set to the value that was
+// read from the store.
+func removeEnvOverrides(cfg, cfgWithoutEnv *model.Config, envOverrides map[string]interface{}) *model.Config {
+	paths := getPaths(envOverrides)
+	newCfg := cfg.Clone()
+	for _, path := range paths {
+		currentVal := getVal(newCfg, path)
+		originalVal := getVal(cfgWithoutEnv, path)
+		if currentVal.Interface() != originalVal.Interface() {
```

Comment:
Could we just /always/ `newVal.Set`? Presumably this will almost always be true, unless we happened to override to the existing configuration value, which seems unlikely.

Reply:
...Right, I think you're right. The only time we would _not_ want to replace a config value, is if the user changed the config value manually. And since we don't allow them to change config values that are overridden by the environment, that can't happen. Correct?

Q: why don't we allow them to change variables that have been set through the environment? Are MM instances running for weeks/months/years? If so, it seems like it's a bit of an inconvenience, no?

Reply:
Conceivably, we could allow a temporary change to the configuration, such that we discard the environment override if set via the system console, but it would unexpectedly come back on restart when the variable is overridden again -- there's no way to know to discard it then.

Environment overrides are primarily intended to be used to tweak server-specific settings -- e.g. changing SiteURL on community-daily.mattermost.com vs community.mattermost.com. They were also used for Docker-like environments, where we also used to have the read-only config, but now that we have the database config, I expect them to be used a lot less frequently.